internal/core/src/common/Array.h

```diff
@@ -0,0 +1,639 @@
+                    if (get_data<int64_t>(i) != arr.get_data<int64_t>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            case DataType::BOOL: {
+                for (int i = 0; i < length_; ++i) {
+                    if (get_data<bool>(i) != arr.get_data<bool>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            case DataType::DOUBLE: {
+                for (int i = 0; i < length_; ++i) {
+                    if (get_data<double>(i) != arr.get_data<double>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            case DataType::FLOAT: {
+                for (int i = 0; i < length_; ++i) {
+                    if (get_data<float>(i) != arr.get_data<float>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            case DataType::INT32:
+            case DataType::INT16:
+            case DataType::INT8: {
+                for (int i = 0; i < length_; ++i) {
+                    if (get_data<int>(i) != arr.get_data<int>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            case DataType::STRING:
+            case DataType::VARCHAR: {
+                for (int i = 0; i < length_; ++i) {
+                    if (get_data<std::string_view>(i) !=
+                        arr.get_data<std::string_view>(i)) {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            default:
+                PanicCodeInfo(Unsupported,
+                              "unsupported element type for array");
+        }
+    }
+
+    template <typename T>
+    T
+    get_data(const int index) const {
+        AssertInfo(
+            index >= 0 && index < length_,
+            fmt::format(
+                "index out of range, index={}, length={}", index, length_));
+        size_t element_length = (index == length_ - 1)
+                                    ? size_ - offsets_.back()
+                                    : offsets_[index + 1] - offsets_[index];
+        if constexpr (std::is_same_v<T, std::string> ||
+                      std::is_same_v<T, std::string_view>) {
+            return T(data_ + offsets_[index], element_length);
+        }
+        if constexpr (std::is_same_v<T, int> || std::is_same_v<T, int64_t> ||
+                      std::is_same_v<T, float> || std::is_same_v<T, double>) {
+            switch (element_type_) {
+                case DataType::INT8:
+                case DataType::INT16:
+                case DataType::INT32:
+                    return static_cast<T>(reinterpret_cast<int*>(data_)[index]);
```

Comment:
reinterpret to `int32_t`, the size of `int` may be 64 bits, not always 32 bits on all platforms.

The other codes like this should be also modified